### PR TITLE
Python3.10

### PR DESCRIPTION
**Pull request by Matthew-Jennings**

### CODE REVIEW

File: .github/workflows/library.yml [hunk 1]
```diff
@@ -534,7 +534,7 @@ jobs:
       fail-fast: false
       matrix:
         os: ["ubuntu", "macos", "windows"]
-        python-version: [3.6, 3.7, 3.8]
+        python-version: [3.6, 3.7, 3.8] #TODO: Does this need updating?
```
Comment (Matthew-Jennings) [inline .github/workflows/library.yml:537]: Does this need updating?

File: requirements-cli.txt [hunk 2]
```diff
@@ -1,2 +1,2 @@
 toml==0.10.2; (python_version >= "2.6" and python_full_version < "3.0.0") or (python_full_version >= "3.3.0")
-typing-extensions==4.0.1; python_version >= "3.6"
+typing-extensions==4.2.0; python_version >= "3.7"
```
Comment (Matthew-Jennings) [inline requirements-cli.txt:2]: And here, or no?

File: .github/workflows/library.yml (hunk 1 above)
Comment (SimonBiggs) [reply .github/workflows/library.yml:537]: It's actually a "dead bit of code" not currently in use... Left over from an attempt to automate the release process.

File: requirements-cli.txt (hunk 2 above)
Comment (SimonBiggs) [reply requirements-cli.txt:2]: Requirements files aren't to be updated manually. They're created from the poetry lock file when propagate is ran.

File: .github/workflows/library.yml (hunk 1 above)
Comment (SimonBiggs) [reply .github/workflows/library.yml:537]: Likely still worth keeping around if we ever do want to automate the release process...

Comment (sjswerdloff) [reply .github/workflows/library.yml:537]: but if we keep it around, shouldn't we update it to the versions of python being supported?
We at least support 3.9, we are looking at supporting 3.10, and without double checking (sorry), I thought we don't support 3.6 and are about to stop supporting 3.7...

**Comment by Matthew-Jennings** [comment]
So I think we have some success!

- `imageio` now seems to work unpinned due to changes in [lib/pymedphys/_losslessjpeg/\_\_init\_\_.py](https://github.com/pymedphys/pymedphys/pull/1648/files#diff-a87fc07bcffa055b1f7aa8757b0261ec0d99bd201b6ee2970c41f267f4c5b133)
- docs build works on Windows with Python 3.10 by replacing `sphinxcontrib-napoleon` with `sphinx.ext.napoleon` in [lib/pymedphys/docs/_config.yml](https://github.com/pymedphys/pymedphys/pull/1648/files#diff-01435d79a70be63926d34c00d261e925753fd462784752ed176ec22b905545fe).
- I can now successfully run `poetry install -E dev` locally using Python 3.10. CI is also successful.

**Review by SimonBiggs — CHANGES_REQUESTED** [review]
Some brilliant work here :)

There seems to be a few CI items that are a little unhappy still though...

File: .github/workflows/library.yml (hunk 1 above)
Comment (Matthew-Jennings) [reply .github/workflows/library.yml]: I edited them as per @sjswerdloff's points.

**Comment by Matthew-Jennings** [comment]
> Some brilliant work here :)
> 
> There seems to be a few CI items that are a little unhappy still though...

Latest changes _might_ do the trick.

- `pylint` appears to be throwing false positive E1101s (no-member) for `pathlib.Path.joinpath()` (see similar [Github issue](https://github.com/PyCQA/pylint/issues/5832)). I think its safe to disable these inline. COuld maybe disable more broadly, but not sure how best to do so.

- `pylint` is suddenly throwing [E0601 (used-before-assignment) errors too. I've put in dummy assignments to satisfy pylint, but wonder whether it is acceptable to just simply disable E0601..

- doctests were failing due to `streamlit` and an incompatibly version of `protobuf`. `streamlit` is pinned and updating it _might_ help and is possibly a better long-term solution but I didn't investigate this due to a warning in `pyproject.toml` indicating the need for its pin.

File: lib/pymedphys/_utilities/test.py [hunk 3]
```diff
@@ -74,15 +74,19 @@ def get_executable_even_when_embedded():
         except FileNotFoundError:
             pass
 
-    exe = str(pathlib.Path(np.__file__).parents[4].joinpath("bin", "python"))
+    exe = str(
+        pathlib.Path(np.__file__).parents[4].joinpath("bin", "python")
+    )  # pylint: disable=no-member
```
Comment (SimonBiggs) [inline lib/pymedphys/_utilities/test.py:79]: This need for no-member has me confused...

Comment (Matthew-Jennings) [reply lib/pymedphys/_utilities/test.py:79]: One possibility is that `pylint` somehow recognises that `pathlib.Path(np.__file__).parents[4]` could return an Exception (IndexError) and so `joinpath()` would not be a "member". Dunno.

**Comment by Matthew-Jennings** [comment]
I think I got everything except the Mosaiq tests passing (notwithstanding that not every combination of test task, python version and OS are being tested!). 

I'm likely not going to be able to spend any longer on this for a couple of days, but understand that:
- the Mosaiq tests may need addressing before the merge
- we'll need a changelog entry.

I'll try to revisit midweek. In the meantime, if anyone (@SimonBiggs, @dg1an3?) has a quick fix for the Mosaiq stuff they can add, I'd be very much obliged!

**Comment by SimonBiggs** [comment]
> * `pylint` is suddenly throwing [E0601 (used-before-assignment) errors too. I've put in dummy assignments to satisfy pylint, but wonder whether it is acceptable to just simply disable E0601..

My rule of thumb, if a new pylint error is added I will initially add it to the list of ignored errors in the config file, and then at a later date decide whether or not to scour the repo and fix up all of those.

> * doctests were failing due to `streamlit` and an incompatibly version of `protobuf`. `streamlit` is pinned and updating it _might_ help and is possibly a better long-term solution but I didn't investigate this due to a warning in `pyproject.toml` indicating the need for its pin.

Streamlit is okay to be upgraded, it just needs to stay pinned for end user installation for now. But, streamlit is stabalising significantly. So this might not be an issue. But if upgrading streamlit doesn't cause CI issues, it's fair game.

**Comment by SimonBiggs** [comment]
> But if upgrading streamlit doesn't cause CI issues, it's fair game.

According to the Beyonce rule... if the feature isn't covered by tests then:

"If you liked it you shoulda put CI on it" ...

**Comment by SimonBiggs** [comment]
The upgraded pandas is clashing with the currently installed sqlalchemy version:

```
E           ImportError: Unable to find a usable engine; tried using: 'sqlalchemy'.
E           A suitable version of sqlalchemy is required for sql I/O support.
E           Trying to import the above resulted in these errors:
E            - Pandas requires version '1.4.0' or newer of 'sqlalchemy' (version '1.3.24' currently installed).
```

I seem to remember that upgrading sqlalchemy broke @dg1an3's tests... so I opted to pin sqlalchemy to keep the tests working.

File: pyproject.toml [hunk 4]
```diff
@@ -31,93 +31,88 @@ packages = [{ include = "pymedphys", from = "lib" }]
...
+watchdog = { version = "*", optional = true } # groups = ["user", "dev"]
+
+keyring = { version = "*", optional = true }         # groups = ["user", "dev"]
+packaging = { version = "*", optional = true }       # groups = ["user", "dev"]
+PyYAML = { version = "*", optional = true }          # groups = ["user", "dev"]
+requests = { version = "*", optional = true }        # groups = ["user", "dev"]
+python-dateutil = { version = "*", optional = true } # groups = ["user", "dev", "tests"]
+
+matplotlib = { version = "*", optional = true } # groups = ["user", "dev"]
+scipy = "1.8.1"                                 # groups = ["user", "dev"]
```
Comment (SimonBiggs) [inline pyproject.toml:52]: I'm assuming we don't want this pin left here? (A left over from debugging?)

**Comment by SimonBiggs** [comment]
Yeah @dg1an3,

It might be quite awesome to be able to have your help here. I while back I noticed that the Mosaiq tests written ended up failing if sqlalchemy was upgraded to version 1.4 or above. Do you have any ideas about how best to approach this?

File: pyproject.toml (hunk 4 above)
Comment (Matthew-Jennings) [reply pyproject.toml]: No sorry - I could've sworn I'd corrected that? Maybe a relic of incorrectly merging auto-changes (propagate?) from Github.

**Comment by Matthew-Jennings** [comment]
> The upgraded pandas is clashing with the currently installed sqlalchemy version:
> 
> ```
> E           ImportError: Unable to find a usable engine; tried using: 'sqlalchemy'.
> E           A suitable version of sqlalchemy is required for sql I/O support.
> E           Trying to import the above resulted in these errors:
> E            - Pandas requires version '1.4.0' or newer of 'sqlalchemy' (version '1.3.24' currently installed).
> ```
> 
> I seem to remember that upgrading sqlalchemy broke @dg1an3's tests... so I opted to pin sqlalchemy to keep the tests working.

I'm sorry for the messy unpins. I honestly thought I'd reverted the unpins back once I'd diagnosed the issue, but apparently I missed `sqlalchemy` either in that effort or in the throes of merging Github-generated commits

**Comment by SimonBiggs** [comment]
> possibly due to Python3.9?

Not actually due to 3.9. @Matthew-Jennings actually did quite a bit of work in here fixing the various tests by updating the required dependencies.

**Comment by Matthew-Jennings** [comment]
> > Or @nlambriICH, might you be interested in taking a swing at this?
> 
> @SimonBiggs I could give it a try, but I think someone more experienced than me might be more suitable.
> 
> As a sidenote, I see that here the tests _Library / Tests (ubuntu, updates) (pull_request)_ and _Library / Tests (windows, docs) (pull_request)_ are passing, whereas in #1638 they are failing, possibly due to Python3.9?

Hi @nlambriICH, if you're up for it, you could have a go! I'm likely out until next week now - I won't have access to a suitable workstation for coding. But I can still review PRs, etc.

I think the way to go would be to:

1. create a new branch/pull request from _this_ branch (`python3.10`)
2. try removing the `sqlalchemy` pin in `pyproject.toml` and run `poetry update` and `poetry install -E dev`. You may need to specifically run `poetry update sqlalchemy`
3. Push to Github and see whether the Mosaiq tests fail and, if so, begin troubleshooting
   - if it does fail, it may be because of a different reason that it's failing now. I.e. it may fail because upgrading `sqlalchemy` broke the Mosaiq database test code, as opposed to whatever reason the test is failing now (which upgrading `sqlalchemy` may fix!)
   - If upgrading `sqlalchemy` does make it fail (differently), query the `sqlalchemy` release notes. They have introduced some relevant breaking changes between our currently pinned version and the latest version. See if you can identify them - with luck it might be fairly simple - e.g. a change in function name/parameters?

Don't worry about all of that if you're not keen. I only write this since you expressed interest in trying!

**Comment by Matthew-Jennings** [comment]
> Hi @nlambriICH, if you're up for it, you could have a go! I'm likely out until next week now - I won't have access to a suitable workstation for coding. But I can still review PRs, etc.
> 
> I think the way to go would be to:
> 
> 1. create a new branch/pull request from _this_ branch (`python3.10`)
> 2. try removing the `sqlalchemy` pin in `pyproject.toml` and run `poetry update` and `poetry install -E dev`. You may need to specifically run `poetry update sqlalchemy`
> 3. Push to Github and see whether the Mosaiq tests fail and, if so, begin troubleshooting
>    
>    * if it does fail, it may be because of a different reason that it's failing now. I.e. it may fail because upgrading `sqlalchemy` broke the Mosaiq database test code, as opposed to whatever reason the test is failing now (which upgrading `sqlalchemy` may fix!)
>    * If upgrading `sqlalchemy` does make it fail (differently), have a look at the `sqlalchemy` release notes. They may have introduced some relevant breaking changes between our currently pinned version and the latest version. See if you can identify a culprit - with luck it might be fairly simple - e.g. a change in function name/parameters?
> 
> Don't worry about all of that if you're not keen. I only write this since you expressed interest in trying!

I had a small stab at this to no avail. Though my SQL skills are little more than "beginner" level.

**Comment by SimonBiggs** [comment]
> I had a small stab at this to no avail. Though my SQL skills are little more than "beginner" level.

Alright, having these tests not pass is blocking other PRs being able to be merged. I'll give this a crack.

**Comment by SimonBiggs** [comment]
Given this is merging into the other Python 3.10 branch, I'll merge this into there and work on it further there.

**Pull request merged**